From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 20 Aug 2024 17:32:40 +0400
Subject: [PATCH] Paper PR: Prevent zombie reinforcements loading chunks


diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 9460ffdede40a2e3601d3c97b1d1ca4e62dcbf29..9de2ba661b75ff6b819cb8361baa0d59870fd139 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -341,6 +341,12 @@ public class Zombie extends Monster {
                     BlockPos blockposition = new BlockPos(i1, j1, k1);
                     EntityType<?> entitytypes = entityzombie.getType();
 
+                    // Paper start - Prevent reinforcement checks from loading chunks
+                    if (this.level().getChunkIfLoadedImmediately(blockposition.getX() >> 4, blockposition.getZ() >> 4) == null) {
+                        continue;
+                    }
+                    // Paper end - Prevent reinforcement checks from loading chunks
+
                     if (SpawnPlacements.isSpawnPositionOk(entitytypes, this.level(), blockposition) && SpawnPlacements.checkSpawnRules(entitytypes, worldserver, MobSpawnType.REINFORCEMENT, blockposition, this.level().random)) {
                         entityzombie.setPos((double) i1, (double) j1, (double) k1);
                         if (!this.level().hasNearbyAlivePlayerThatAffectsSpawning((double) i1, (double) j1, (double) k1, 7.0D) && this.level().isUnobstructed(entityzombie) && this.level().noCollision((Entity) entityzombie) && !this.level().containsAnyLiquid(entityzombie.getBoundingBox())) { // Paper - Affects Spawning API
